lib/howl/interactions/replacement.moon

```diff
@@ -0,0 +1,385 @@
+-- Copyright 2012-2015 The Howl Developers
+-- License: MIT (see LICENSE.md at the top-level directory of the distribution)
+
+import Matcher, TaskRunner from howl.util
+import app, command, config, interact, timer from howl
+import highlight, ListWidget, NotificationWidget from howl.ui
+
+append = table.insert
+
+highlight.define_default 'replace_strikeout', {
+  style: highlight.STRIKE,
+  color: '#f0f0f0'
+}
+
+parse_replacement = (text, marker) ->
+  target, replacement = text\match("([^#{marker}]+)#{marker}(.*)")
+  if target
+    return target, replacement
+  return text
+
+class Replacement
+  run: (@finish, opts={}) =>
+    @command_line = app.window.command_line
+    with opts.editor
+      @text = .buffer.text
+      @line_at_top = .line_at_top
+      @orig_cursor_pos = .cursor.pos
+
+    -- @preview_buffer holds replaced text
+    @preview_buffer = app\new_buffer opts.editor.buffer.mode
+    @preview_buffer.title = opts.title or 'Preview Replacements'
+    @preview_buffer.text = @text
+
+    -- @buffer always holds original text
+    @buffer = howl.Buffer!
+    @buffer.text = @text
+
+    @start_pos = opts.editor.active_chunk.start_pos or 1
+    @end_pos = opts.editor.active_chunk.end_pos or @buffer.length
+
+    @runner = TaskRunner!
+
+    self.find = opts.find or (buffer, target, init) ->
+      init = buffer\byte_offset init
+      text = buffer.text
+      ->
+        start_pos, end_pos  = text\find target, init, true
+        if not start_pos
+          return
+        init = end_pos + 1
+        return { start_pos: buffer\char_offset(start_pos), end_pos: buffer\char_offset(end_pos) }
+
+    self.replace = opts.replace or (match, replacement) -> replacement
+
+    @orig_buffer = app.editor.buffer
+    app.editor.buffer = @preview_buffer
+    app.editor.line_at_top = @line_at_top
+
+    @caption_widget = NotificationWidget!
+    @command_line\add_widget 'caption_widget', @caption_widget
+
+    @selected_idx = nil
+    @num_matches = 0
+    @num_replacements = 0
+    @matches = {}
+    @replacements = {}
+    @replacements_applied = {}
+    @strikeouts = {}
+    @excluded = {}
+    @num_excluded = 0
+    @adjusted_positions = {}
+
+    replace_command = (opts.replace_command or '') .. (@command_line\pop_spillover! or '')
+    if replace_command.is_empty
+      replace_command = '/'
+    @command_line\write replace_command
+
+    @on_update replace_command
+
+  _restore_return: (result) =>
+    @runner\cancel_all!
+    app.editor.buffer = @orig_buffer
+    app\close_buffer @preview_buffer, true
+    self.finish result
+    true
+
+  on_update: (text) =>
+    return if not text or text.is_empty
+
+    marker = text[1]
+    target, replacement = parse_replacement text\usub(2), marker
+
+    local loader
+    if target != @target
```

Comment:
I just had to test this and find the case that doesn't work :)

I entered a target and replacement, and then when back and updated the target. This reloads the matches, which removes any replacements, but since the replacement wasn't changed these aren't reloaded.